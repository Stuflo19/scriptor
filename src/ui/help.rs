use ratatui::{
    style::{Modifier, Style, Stylize},
    text::{Line, Text},
    widgets::Paragraph,
};

pub fn generate_help_text() -> Paragraph<'static> {
    let (msg, style) = (
        vec!["Press ".into(), "esc".bold(), " to exit, ".into()],
        Style::default().add_modifier(Modifier::RAPID_BLINK),
    );
    let text = Text::from(Line::from(msg)).patch_style(style);
    
    Paragraph::new(text)
}
