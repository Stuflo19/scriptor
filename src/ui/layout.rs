use ratatui::layout::{Constraint, Layout, Rect};

pub fn generate_layout(frame_area: Rect) -> [Rect; 3] {
    let vertical = Layout::vertical([
        Constraint::Length(1),
        Constraint::Length(3),
        Constraint::Min(1),
    ]);

    vertical.areas(frame_area)
}
