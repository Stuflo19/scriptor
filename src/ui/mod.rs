pub mod help;
pub mod input;
pub mod layout;
pub mod render;
pub mod table;
