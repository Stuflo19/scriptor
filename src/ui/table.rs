use crate::file_reader::scripts::Scripts;
use ratatui::widgets::TableState;
use ratatui::{
    layout::Constraint,
    style::{Style, Stylize},
    text::{Span, Text},
    widgets::{Block, Cell, Row, Table},
};
use std::collections::HashMap;

pub struct ScriptTable {
    pub state: TableState,
    pub scripts: Scripts,
    pub filtered: HashMap<String, String>,
}

impl ScriptTable {
    pub fn new() -> Self {
        let scripts = Scripts::new();
        let temp = scripts.scripts.clone();

        Self {
            state: TableState::default().with_selected(0),
            scripts,
            filtered: temp,
        }
    }

    fn generate_table_header(&self) -> Row<'static> {
        let header_style = Style::default().on_black();

        ["Script", "Command"]
            .into_iter()
            .map(Cell::from)
            .collect::<Row>()
            .style(header_style)
            .height(1)
    }

    pub fn next(&mut self) {
        let i = match self.state.selected() {
            Some(i) => {
                if i >= self.scripts.scripts.len() - 1 {
                    0
                } else {
                    i + 1
                }
            }
            None => 0,
        };
        self.state.select(Some(i));
        //self.scroll_state = self.scroll_state.position(i * ITEM_HEIGHT);
    }

    pub fn previous(&mut self) {
        let i = match self.state.selected() {
            Some(i) => {
                if i == 0 {
                    self.scripts.scripts.len() - 1
                } else {
                    i - 1
                }
            }
            None => 0,
        };
        self.state.select(Some(i));
        //self.scroll_state = self.scroll_state.position(i * ITEM_HEIGHT);
    }

    pub fn filter_results(&mut self, input: String) {
        self.filtered = self
            .scripts
            .scripts
            .clone()
            .into_iter()
            .filter(|(key, _value)| key.contains(&input))
            .collect();
    }

    pub fn get_script_name(&self) -> String {
        let index = self.state.selected().unwrap();

        let keys: Vec<String> = self.scripts.scripts.clone().into_keys().collect();

        keys[index].clone()
    }

    fn generate_table_rows(&self) -> Vec<Row<'static>> {
        let script_rows = self
            .filtered
            .iter()
            .enumerate()
            .map(|(_i, (key, value))| {
                let content = [
                    Cell::from(Span::raw(format!("{key}"))),
                    Cell::from(Span::raw(format!("{value}"))),
                ];

                Row::new(content)
            })
            .collect::<Vec<Row>>();

        script_rows
    }

    pub fn generate_table(&self) -> Table<'static> {
        let bar = " â–ˆ ";

        let header = self.generate_table_header();

        let script_rows = self.generate_table_rows();

        let scripts_table = Table::new(
            script_rows,
            [
                Constraint::Length(25 + 1),
                Constraint::Min(25 + 1),
                Constraint::Min(25),
            ],
        )
        .header(header)
        .block(Block::bordered())
        .highlight_symbol(
            Text::from(vec![bar.into(), bar.into(), bar.into(), bar.into()])
                .style(Style::new().red()),
        );

        scripts_table
    }
}
