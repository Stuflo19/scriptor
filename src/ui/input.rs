use ratatui::{
    layout::{Position, Rect},
    style::{Color, Style},
    widgets::{Block, Paragraph},
};

pub struct ScriptInput {
    pub input: String,
    character_index: usize,
}

impl ScriptInput {
    pub fn new() -> Self {
        Self {
            input: String::new(),
            character_index: 0,
        }
    }

    fn clamp_cursor(&self, new_cursor_pos: usize) -> usize {
        new_cursor_pos.clamp(0, self.input.chars().count())
    }

    pub fn move_cursor_left(&mut self) {
        let cursor_moved_left = self.character_index.saturating_sub(1);
        self.character_index = self.clamp_cursor(cursor_moved_left);
    }

    pub fn move_cursor_right(&mut self) {
        let cursor_moved_right = self.character_index.saturating_add(1);
        self.character_index = self.clamp_cursor(cursor_moved_right);
    }

    pub fn place_cursor(&self, input_area: Rect) -> Position {
        Position::new(
            // Draw the cursor at the current position in the input field.
            // This position is can be controlled via the left and right arrow key
            input_area.x + self.character_index as u16 + 1,
            // Move one line down, from the border to the input line
            input_area.y + 1,
        )
    }

    pub fn enter_char(&mut self, new_char: char) {
        let index = self.byte_index();
        self.input.insert(index, new_char);
        self.move_cursor_right();

        // self.filter_results();
        // self.state.select(Some(0));
    }

    /// Returns the byte index based on the character position.
    ///
    /// Since each character in a string can contain multiple bytes, it's necessary to calculate
    /// the byte index based on the index of the character.
    fn byte_index(&self) -> usize {
        self.input
            .char_indices()
            .map(|(i, _)| i)
            .nth(self.character_index)
            .unwrap_or(self.input.len())
    }

    pub fn delete_char(&mut self) {
        let is_not_cursor_leftmost = self.character_index != 0;
        if is_not_cursor_leftmost {
            // Method "remove" is not used on the saved text for deleting the selected char.
            // Reason: Using remove on String works on bytes instead of the chars.
            // Using remove would require special care because of char boundaries.

            let current_index = self.character_index;
            let from_left_to_current_index = current_index - 1;

            // Getting all characters before the selected character.
            let before_char_to_delete = self.input.chars().take(from_left_to_current_index);
            // Getting all characters after selected character.
            let after_char_to_delete = self.input.chars().skip(current_index);

            // Put all characters together except the selected one.
            // By leaving the selected one out, it is forgotten and therefore deleted.
            self.input = before_char_to_delete.chain(after_char_to_delete).collect();
            self.move_cursor_left();

            // self.filter_results();
            // self.state.select(Some(0));
        }
    }

    pub fn generate_input(&self) -> Paragraph<'static> {
        Paragraph::new(self.input.clone())
            .style(Style::default().fg(Color::Yellow))
            .block(Block::bordered().title("Search"))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ui::layout::generate_layout;
    use insta::assert_snapshot;
    use ratatui::{backend::TestBackend, Terminal};

    #[test]
    fn it_new_creates_default_state() {
        let script_input = ScriptInput::new();

        assert_eq!(script_input.input, "");
        assert_eq!(script_input.character_index, 0);
    }

    #[test]
    fn it_move_left_correctly_updates_state() {
        let mut script_input = ScriptInput::new();
        script_input.input = String::from("abc");
        script_input.character_index = 3;

        script_input.move_cursor_left();

        assert_eq!(script_input.character_index, 2);
    }

    #[test]
    fn it_move_left_correctly_handles_0_character_index() {
        let mut script_input = ScriptInput::new();

        script_input.move_cursor_left();

        assert_eq!(script_input.character_index, 0);
    }

    #[test]
    fn it_move_right_correctly_updates_state() {
        let mut script_input = ScriptInput::new();
        script_input.input = String::from("abc");
        script_input.character_index = 2;

        script_input.move_cursor_right();

        assert_eq!(script_input.character_index, 3);
    }

    #[test]
    fn it_move_right_correctly_handles_going_past_input_len() {
        let mut script_input = ScriptInput::new();
        script_input.input = String::from("abc");
        script_input.character_index = 3;

        script_input.move_cursor_right();

        assert_eq!(script_input.character_index, 3);
    }

    #[test]
    fn it_enter_char_correctly_updates_state() {
        let mut script_input = ScriptInput::new();

        script_input.enter_char('a');

        assert_eq!(script_input.input, "a");
        assert_eq!(script_input.character_index, 1);
    }

    #[test]
    fn it_delete_char_correctly_updates_state() {
        let mut script_input = ScriptInput::new();
        script_input.input = String::from("abc");
        script_input.character_index = 3;

        script_input.delete_char();

        assert_eq!(script_input.input, "ab");
        assert_eq!(script_input.character_index, 2);
    }

    #[test]
    fn it_delete_char_handles_empty_input() {
        let mut script_input = ScriptInput::new();

        script_input.delete_char();

        assert_eq!(script_input.input, "");
        assert_eq!(script_input.character_index, 0);
    }

    #[test]
    fn it_generate_input_renders_correctly() {
        let script_input = ScriptInput::new();
        let input = script_input.generate_input();

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();

        terminal
            .draw(|frame| {
                let [_, input_area, _] = generate_layout(frame.area());

                frame.set_cursor_position(script_input.place_cursor(input_area));
                frame.render_widget(&input, input_area)
            })
            .unwrap();

        assert_snapshot!(terminal.backend());
    }
}
