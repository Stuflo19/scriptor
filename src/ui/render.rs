use crate::ui;
use crate::ui::input::ScriptInput;
use crate::ui::table::ScriptTable;
use crossterm::event;
use crossterm::event::{Event, KeyCode};
use ratatui::{DefaultTerminal, Frame};

pub struct Render {
    script_table: ScriptTable,
    script_input: ScriptInput,
}

impl Render {
    pub fn new() -> Self {
        let script_table = ScriptTable::new();
        let script_input = ScriptInput::new();

        Self {
            script_table,
            script_input,
        }
    }

    pub fn run(&mut self, mut terminal: DefaultTerminal) -> color_eyre::Result<String> {
        loop {
            terminal.draw(|frame| self.draw(frame))?;

            if let Event::Key(key) = event::read()? {
                match key.code {
                    KeyCode::Enter => {
                        if !self.script_table.scripts.scripts.is_empty() {
                            let script_name = self.script_table.get_script_name();
                            return Ok(script_name);
                        }
                    }
                    KeyCode::Char(to_insert) => {
                        self.script_input.enter_char(to_insert);
                        self.script_table
                            .filter_results(self.script_input.input.clone());
                    }
                    KeyCode::Backspace => {
                        let curr_input = self.script_input.input.clone();

                        self.script_input.delete_char();
                        if !curr_input.is_empty() {
                            self.script_table
                                .filter_results(self.script_input.input.clone());
                        }
                    }
                    KeyCode::Left => self.script_input.move_cursor_left(),
                    KeyCode::Right => self.script_input.move_cursor_right(),
                    KeyCode::Esc => return Ok("none".to_string()),
                    KeyCode::Up => self.script_table.previous(),
                    KeyCode::Down => self.script_table.next(),
                    _ => {}
                }
            }
        }
    }

    pub fn draw(&self, frame: &mut Frame) {
        let [help_area, input_area, messages_area] = ui::layout::generate_layout(frame.area());

        let input_field = self.script_input.generate_input();

        frame.render_widget(ui::help::generate_help_text(), help_area);

        frame.render_widget(input_field, input_area);

        frame.set_cursor_position(self.script_input.place_cursor(input_area));

        frame.render_stateful_widget(
            self.script_table.generate_table(),
            messages_area,
            &mut self.script_table.state.clone(),
        );
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use ratatui::backend::TestBackend;
    use ratatui::Terminal;

    #[test]
    fn it_draw_snapshot() {
        let render = Render::new();
        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();

        terminal
            .draw(|frame| {
                render.draw(frame);
            })
            .unwrap();

        assert_snapshot!(terminal.backend());
    }
}
