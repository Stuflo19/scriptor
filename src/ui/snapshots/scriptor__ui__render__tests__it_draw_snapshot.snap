---
source: src/ui/render.rs
expression: terminal.backend()
snapshot_kind: text
---
"Press esc to exit,                                                              "
"┌Search────────────────────────────────────────────────────────────────────────┐"
"│                                                                              │"
"└──────────────────────────────────────────────────────────────────────────────┘"
"┌Scripts───────────────────────────────────────────────────────────────────────┐"
"│   Script                 Command                                             │"
"│ █ android                yarn android                                        │"
"│   ios                    yarn ios                                            │"
"│   run                    expo start                                          │"
"│   start                  yarn start                                          │"
"│   test                   jest                                                │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"│                                                                              │"
"└──────────────────────────────────────────────────────────────────────────────┘"
